browserdebuggertools/wssessionmanager.py

```diff
@@ -97,86 +173,102 @@ def __init__(self, port, timeout, domains=None):
-        try:
-            self.close()
-        except:
-            pass
+        self.close()
 
-    def _setup_websocket(self):
-        logging.info("Connecting to websocket %s" % self._websocket_url)
-        self._websocket = websocket.create_connection(
-            self._websocket_url, timeout=self.CONN_TIMEOUT
-        )
-        self._websocket.settimeout(0)  # Don"t wait for new messages
+    def _check_messaging_thread(self):
 
-        for domain, params in self._domains.items():
-            self.enable_domain(domain, params)
+        if self.messaging_thread.is_alive():
+            if self.messaging_thread.blocked:
+                logging.warning("WS messaging thread appears to be blocked")
+                self.close()
+                restart = True
+            else:
+                restart = False
+        else:
+            if isinstance(self.messaging_thread.exception,
+                          websocket.WebSocketConnectionClosedException):
+                restart = True
+            elif self.messaging_thread.exception:
+                raise self.messaging_thread.exception
+            else:
+                raise DeadMessagingThread("WS messaging thread died for an unknown reason")
 
-        return self._websocket
+        if restart:
+            self.increment_messaging_thread_not_ok()
+            self.setup_ws_session()
 
-    def increment_connection_closed_count(self):
+    def increment_messaging_thread_not_ok(self):
 
         now = datetime.now()
 
         if (
-                self._connection_last_closed and
-                (now - self._connection_last_closed).seconds > self.RETRY_COUNT_TIMEOUT
+            self._last_not_ok and
+            (now - self._last_not_ok).seconds > self.RETRY_COUNT_TIMEOUT
         ):
-            self._connection_closed_count = 0
+            self._messaging_thread_not_ok_count = 0
 
-        self._connection_last_closed = now
-        self._connection_closed_count += 1
+        self._last_not_ok = now
+        self._messaging_thread_not_ok_count += 1
 
-        if self._connection_closed_count > self.MAX_CONNECTION_RETRIES:
+        if self._messaging_thread_not_ok_count > self.MAX_RETRY_THREADS:
             raise MaxRetriesException(
-                "Websocket connection found closed %s times within %s seconds" % (
-                    self.MAX_CONNECTION_RETRIES, self.RETRY_COUNT_TIMEOUT
+                "WS messaging thread not ok %s times within %s seconds" % (
+                    self.MAX_RETRY_THREADS, self.RETRY_COUNT_TIMEOUT
                 )
             )
 
-        self._setup_websocket()
+    def setup_ws_session(self):
+
+        self.messaging_thread = _WSMessagingThread(
+            self.port, self._send_queue, self._recv_queue, self._poll_signal
+        )
+        self.messaging_thread.start()
+
+        for domain, params in self._domains.items():
+            self.enable_domain(domain, params)
+        return self.messaging_thread
```

Comment:
I don't think you need to return `self.messaging_thread` here

Reply:
This is so we can declare in the constructor
`self.messaging_thread = self.setup_ws_session()`

Reply:
My point was that `setup_ws_session` sets it anyway as a side effect (and you also for instance use it this way when you restart).

It is a minor point, of course, and I guess one mainly concerning style. I just think it's a bit redundant (if the method sets a variable as a side effect then it does not need to return it: if you prefer declaring `self.messaging_thread` in the constructor so it is more clear you have one, then I would expect something like `self.messaging_thread = None` before you call the setup).

I'll leave it to you to decide

Reply:
ok sure, i'll go with `self.messaging_thread = None`, thanks.